### PR TITLE
fix geojson position

### PR DESCRIPTION
**Pull request by apiel**
If changing an existing definition:
- Provide a URL to documentation or source code which provides context for the suggested changes: https://geojson.org/schema/GeoJSON.json

In the types definition, we have a coordinate being a `Position`:
```ts
export interface Point extends GeoJsonObject {
    type: "Point";
    coordinates: Position;
}
```
In the official JSON schema definition of GeoJSON (https://geojson.org/schema/GeoJSON.json), we have the coordinate that is an array of at least 2 numbers:
```json
"coordinates": {
  "type": "array",
  "minItems": 2,
  "items": {
     "type": "number"
  }
},
```

### CODE REVIEW

File: types/geojson/index.d.ts
Comment (danvk) [inline types/geojson/index.d.ts:41]: If the comment above is to be believed, then shouldn't this be `[number, number, number?]`? (i.e. exactly two or three numbers -- equivalent to the commented-out type.)

This change has been proposed and rejected before as being too drastic, see https://github.com/DefinitelyTyped/DefinitelyTyped/pull/21590#discussion_r151582350 and cc @DenisCarriere and @JeffJacobson. I tend to agree -- while the proposed type is more precise, this will be almost certainly be a breaking change for all users of these types, and I'm not sure the added safety is really worth the cost. I patched this into my main work project just now, for example, and got 35 type errors.

Comment (icholy) [reply types/geojson/index.d.ts:41]: @danvk `[number, number, number?]` would accept `[0, 0, undefined]`

**Comment by icholy** [comment]
@apiel do you have a specific use-case you need this for?

File: types/geojson/index.d.ts
Comment (JeffJacobson) [reply types/geojson/index.d.ts:41]: Don't wait for my approval; whatever the rest of you decide is fine. It's been a very long time since I've needed to use the GeoJSON type definition (which was several versions of TypeScript ago), so I don't know what is best at the moment.

In retrospect, I regret changing the version number major to 7946 to match the RFC number. In [semantic versioning](https://semver.org/), if there's a breaking change you're supposed to change the major portion of the version number, but now if you do that it won't match the RFC number.

Is there some way in TypeScript to support the both the more restrictive and less restrictive types, but mark the less restrictive one as being deprecated? So a user using more than three coordinates in a position would merely see a warning, but their code wouldn't break?

Alternatively, could an additional alternative type definition be provided (called something like "geojson-strict") with the more restrictive position definition?

Below is an excerpt from [RFC 7946 Section 3.1.1](https://datatracker.ietf.org/doc/html/rfc7946#section-3.1.1), which states the restriction to either two or three elements for easier reference. (I bolded some of the text which specifies the number of allowed elements.)

>    A position is an array of numbers.  **There MUST be two or more
>    elements.**  The first two elements are longitude and latitude, or
>    easting and northing, precisely in that order and using decimal
>    numbers.  Altitude or elevation MAY be included as an optional third
>    element.
> 
>    **Implementations SHOULD NOT extend positions beyond three elements**
>    because the semantics of extra elements are unspecified and
>    ambiguous.  Historically, some implementations have used a fourth
>    element to carry a linear referencing measure (sometimes denoted as
>    "M") or a numerical timestamp, but in most situations a parser will
>    not be able to properly interpret these values.  The interpretation
>    and meaning of additional elements is beyond the scope of this
>    specification, and additional elements MAY be ignored by parsers.

Comment (JeffJacobson) [reply types/geojson/index.d.ts:41]: > @danvk `[number, number, number?]` would accept `[0, 0, undefined]`

Would it also support an array of only two numbers?

Comment (icholy) [reply types/geojson/index.d.ts:41]: @JeffJacobson yes, but my point is that `[0, 0, undefined]` should **not** be a valid `Position`.

Comment (JeffJacobson) [reply types/geojson/index.d.ts:41]: > @JeffJacobson yes, but my point is that `[0, 0, undefined]` should **not** be a valid `Position`.

Thanks for clarifying. I misunderstood.

Comment (danvk) [reply types/geojson/index.d.ts:41]: If you set `exactOptionalPropertyTypes` then `[0, 0, undefined]` is not assignable to that type ([playground][1]):

```ts
type Position = [number, number, number?];

const p: Position = [1, 2, undefined];
//    ~ Type 'undefined' is not assignable to type 'number'.
```

[1]: https://www.typescriptlang.org/play?exactOptionalPropertyTypes=true#code/C4TwDgpgBACg9gZwJbCXAdlAvFA2ugVwFsAjCAJwBopDSLrazyB+AXQG4AoTgYwwWBQwALliIUaTDlwBGagCZqBdABMIAMyToIKjpwD0+qMagA-KABVw0AOTK1m7SptQkCGnEEBDBMgDm6F4kADbQwHBQoJBQNowUNgB0nEA

Comment (icholy) [reply types/geojson/index.d.ts:41]: That's good to know. However, considering that most people don't have `exactOptionalPropertyTypes` enabled, I'd still choose `[number, number] | [number, number, number]` over `[number, number, number?]`

Comment (apiel) [reply types/geojson/index.d.ts:41]: In the proposal i made, it is not as strict as [number, number] | [number, number, number]
It just enforce to have at least 2 number values in the array, as it should be for a coordinate position. Also look at the official Json schema definition from geojson.

Comment (icholy) [reply types/geojson/index.d.ts:41]: @apiel is this change something you personally need? Or is the motivation purely to be more compliant with the spec?

Comment (apiel) [reply types/geojson/index.d.ts:41]: Originally, I needed it when I was using https://www.npmjs.com/package/typescript-json-schema to generate JSON schema for my swagger OpenAPI documentation. Since the types were not consistent, I ended up to download the JSON schema from  https://geojson.org/schema/GeoJSON.json and directly linked them to my doc (so I don't rely on TypeScript anymore to represent my API).

However, TypeScript is meant to be type safe. I still think that a position should have at least 2 values. So would be great to have it right in the types, especially since we have a clear type definition on geojson.org

**Comment by apiel** [comment]
@[icholy](https://github.com/icholy) what do you mean? My aim is just to be able to have the right type definition for GeoJSON TypeScript.

**Comment by icholy** [comment]
That's not a good enough reason to break peoples code.

**Comment by apiel** [comment]
But I don't understand, how can this break the code of people? If they have a position with only 1 value then there code is wrong anyway. Or could you give an example where having only one number in a position make sense?

**Comment by danvk** [comment]
> But I don't understand, how can this break the code of people? If they have a position with only 1 value then there code is wrong anyway. Or could you give an example where having only one number in a position make sense?

There's no disagreement that the types would become more precise / correct with this change. The issue is that there's a whole spectrum between "convenient to work with" and "precise and correct." This change makes the types more precise but harder to work with, and what's unclear is whether the added precision is worth the added pain.

Here's a simple, correct example that works today but would break with these definitions ([playground][1]):

```ts
const point = [0, 0];
const geom: Point = {
  type: 'Point',
  coordinates: point,
// ~~~~~~~~~~ Type 'number[]' is not assignable to type 'Position'.
};
```

That being said, I looked a little closer at the 35 errors this change produces in my work project, and _all_ of them stemmed from Turf.js having an incompatible definition of GeoJSON. But newer versions of Turf.js use `@types/geojson` (see https://github.com/Turfjs/turf/pull/2158). So when I updated turf, all those errors went away.

So maybe this change isn't as painful as I'd thought. I'm shifting my stance from "opposed" to "ambivalent."

[1]: https://www.typescriptlang.org/play?#code/C4TwDgpgBACg9gZwJbCXAdlAvFA2ugVwFsAjCAJwBopDSLqA6J2s83AXXYG4BYAKAgAPMHHLAoSdMAoAzAIYBjaPEniA3vyhaooSAC4oAIhVTDvPtqgK4ogCaS50hAfjJUGcwF9+-a+gTiIqrYeAAM1KHcvhgBUADmEHBELnDBOBoWOuAQBgDkJsC5lJpWNuT26I4QzlBBUsV8APSNUAB+7R2dUAAq2VC5LBQcuRIINHDicgjIcZUkADbQwHBZkP2uKGjouQz8nub8QA

**Review by icholy — CHANGES_REQUESTED** [review]
This isn't worth changing.

**Comment by apiel** [comment]
Actually, can't we just make a major version for breaking change? Shouldn't this be the purpose of semantic versioning?

**Comment by icholy** [comment]
> If we're never allowed to make a breaking change, we should probably put in a comment that says something to the effect of, "We are aware that the GeoJSON spec says there can only be two or three numbers. Please do not change it because..."

@JeffJacobson this would be my preferred approach.

**Pull request closed without merging**
